rfd/0163-vnet.md

```diff
@@ -0,0 +1,971 @@
+
+On MacOS we will find the hosts' default nameservers by parsing `/etc/resolv.conf`.
+
+Here is a flow chart for the above sequence:
+
+```mermaid
+graph TD
+    start(DNS A record query) --> checkIP{IP Assigned?}
+    checkIP -- Yes --> sendIP[Reply with IP]
+    checkIP -- No --> checkProxy{Matches *.proxy-address?}
+    checkProxy -- Yes --> checkAppExists{App Exists?}
+    checkAppExists -- No --> sendNonAnswer[Send Authoritative Non-Answer]
+    checkAppExists -- Yes --> assignIPToApp[Assign Free IP & Send Authoritative Answer]
+    checkProxy -- No --> checkCustomDNS{Matches *.custom-dns-zone?}
+    checkCustomDNS -- Yes --> determineCluster[Determine Cluster for DNS Zone]
+    determineCluster --> queryCluster[Query Cluster for app's spec.public_addr]
+    queryCluster -- Match Found --> assignIPToApp
+    queryCluster -- No Match --> forward[Forward to default nameserver]
+    checkCustomDNS -- No --> forward[Forward to default nameserver]
+```
+
+Here is a seqence diagram for an example VNet session:
+
+```mermaid
+sequenceDiagram
+    participant Client
+    participant DNS_Resolver as DNS Resolver
+    participant VNet_DNS as VNet
+    participant Teleport as teleport.example.com
+    participant App_Service as App Service
+    participant API as API
+
+    Client->>DNS_Resolver: Ask for A record for api.teleport.example.com
+    DNS_Resolver->>DNS_Resolver: Check own cache
+    DNS_Resolver->>VNet_DNS: Ask for A record for api.teleport.example.com
+    VNet_DNS->>VNet_DNS: Check own cache
+    VNet_DNS->>Teleport: Ask if app "api" exists
+    Teleport->>VNet_DNS: Reply "api" exists
+    VNet_DNS->>VNet_DNS: Assign IP 100.64.0.2 to api.teleport.example.com
+    VNet_DNS->>DNS_Resolver: Return A record with IP 100.64.0.2
+    DNS_Resolver->>Client: Return A record with IP 100.64.0.2
+    Client->>VNet_DNS: Open TCP connection with 100.64.0.2
+    VNet_DNS->>Teleport: Proxy connection to cluster
+    Teleport->>App_Service: Proxy connection to app service
+    App_Service->>API: Proxy connection to 127.0.0.1:8888
+```
+
+### IP range
+
+#### IPv4
+
+VNet IP addresses are only ever used locally on a single device for applications
+connecting to the virtual TUN interface.
+Because these IPs do not need to be routable or advertised outside of the client
+machine, this gives us with essentially two options:
+1. Use a subset of one of the private IP ranges `10/8`, `172.16/12`, or `192.168/16`
+   as defined in [RFC 1918](https://datatracker.ietf.org/doc/html/rfc1918).
+2. Use the carrier-grade NAT shared address space `100.64/10` defined in
+   [RFC 6598](https://datatracker.ietf.org/doc/html/rfc6598).
+
+There is some prior art for using the carrier-grade NAT (CGNAT) shared address
+space for our intended purpose, it is already in use by other VPN-replacement
+apps such as [tailscale](https://tailscale.com/) and [Enclave](https://enclave.io/).
+
+VNet will use the CGNAT range by default, but we will make it configurable at
```

Comment:
I don't really understand the concern, you should just not run VNet on the auth or proxy. I think it would be a bit tricky for VNet to specifically detect and block that

Reply:
I would expect it to exist as logic within the proxy, rather than the vnet.  But after considering this further I will resolve this thread.  Any attack of this design could be conducted with a custom DNS server anyways.